File: cloud/functions/schemas.ts
/**
 * Wrack Telemetry — TypeScript event schema validation for Cloud Functions.
 *
 * This module implements runtime validation of telemetry events against the
 * canonical schemas defined in `shared/telemetry-types/schemas/`.
 *
 * It re-exports the shared TypeScript types from `shared/telemetry-types/`
 * and adds Cloud-Function-specific helpers (batch validation, error formatting).
 *
 * Usage:
 *   const { valid, errors } = validateTelemetryEvent(rawEvent);
 *   if (!valid) { return res.status(400).json({ errors }); }
 */

// ---------------------------------------------------------------------------
// Types (mirrored from shared/telemetry-types/typescript/events.ts)
// ---------------------------------------------------------------------------

export type EventSource = 'ev3' | 'rpi' | 'cloud_functions' | 'web' | 'ios';

export type EventType =
  | 'battery_status'
  | 'command_received'
  | 'command_executed'
  | 'device_status'
  | 'error'
  | 'api_request'
  | 'motor_status'
  | 'sensor_reading'
  | 'terrain_scan'
  | 'connection_status';

export interface TelemetryEventEnvelope {
  event_id: string;
  event_type: EventType;
  source: EventSource;
  timestamp: string;
  payload: Record<string, unknown>;
  session_id?: string | null;
  device_id?: string | null;
  version?: string | null;
  tags?: string[] | null;
  user_id?: string | null;
  correlation_id?: string | null;
}

export type BatteryType = 'rechargeable' | 'alkaline' | 'unknown';

export interface BatteryStatusPayload {
  voltage_mv: number;
  voltage_v?: number;
  current_ma?: number;
  percentage: number;
  battery_type?: BatteryType;
  is_critical?: boolean;
}

export type ControllerType = 'ps4' | 'network_remote' | 'unknown';

export interface CommandReceivedPayload {
  command: string;
  params?: Record<string, unknown> | null;
  controller_type?: ControllerType;
  received_at_ms?: number;
}

export interface CommandExecutedPayload {
  command: string;
  success: boolean;
  duration_ms?: number;
  error_message?: string | null;
  params?: Record<string, unknown> | null;
  controller_type?: ControllerType;
}

export type DeviceStatusValue =
  | 'connected'
  | 'disconnected'
  | 'error'
  | 'stalled'
  | 'initializing';

export type DeviceType = 'motor' | 'sensor' | 'controller' | 'unknown';

export interface DeviceStatusPayload {
  device_name: string;
  device_type?: DeviceType;
  port?: string | null;
  status: DeviceStatusValue;
  previous_status?: DeviceStatusValue | 'unknown' | null;
  error_message?: string | null;
}

export interface ErrorPayload {
  error_type: string;
  error_code?: string | null;
  message: string;
  component?: string | null;
  stack_trace?: string | null;
  context?: Record<string, unknown> | null;
}

export type HttpMethod = 'GET' | 'POST' | 'PUT' | 'PATCH' | 'DELETE' | 'OPTIONS';

export interface ApiRequestPayload {
  endpoint: string;
  method?: HttpMethod;
  command?: string | null;
  status_code: number;
  latency_ms: number;
  robot_response_time_ms?: number | null;
  client_ip_hash?: string | null;
  error_message?: string | null;
}

// ---------------------------------------------------------------------------
// Validation infrastructure
// ---------------------------------------------------------------------------

export interface ValidationError {
  field: string;
  message: string;
}

export interface ValidationResult {
  valid: boolean;
  errors: ValidationError[];
}

// ---------------------------------------------------------------------------
// Constants
// ---------------------------------------------------------------------------

export const VALID_SOURCES: readonly EventSource[] = [
  'ev3', 'rpi', 'cloud_functions', 'web', 'ios',
] as const;

export const VALID_EVENT_TYPES: readonly EventType[] = [
  'battery_status', 'command_received', 'command_executed',
  'device_status', 'error', 'api_request',
  'motor_status', 'sensor_reading', 'terrain_scan', 'connection_status',
] as const;

export const P0_EVENT_TYPES: readonly EventType[] = [
  'battery_status', 'command_received', 'command_executed',
  'device_status', 'error', 'api_request',
] as const;

const VALID_DEVICE_STATUSES: readonly DeviceStatusValue[] = [
  'connected', 'disconnected', 'error', 'stalled', 'initializing',
] as const;

const VALID_DEVICE_TYPES: readonly DeviceType[] = [
  'motor', 'sensor', 'controller', 'unknown',
] as const;

const VALID_BATTERY_TYPES: readonly BatteryType[] = [
  'rechargeable', 'alkaline', 'unknown',
] as const;

const VALID_CONTROLLER_TYPES: readonly ControllerType[] = [
  'ps4', 'network_remote', 'unknown',
] as const;

const VALID_HTTP_METHODS: readonly HttpMethod[] = [
  'GET', 'POST', 'PUT', 'PATCH', 'DELETE', 'OPTIONS',
] as const;

const ISO_8601_RE = /^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}(\.\d+)?Z$/;
const UUID_RE = /^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/i;

// ---------------------------------------------------------------------------
// Envelope validation
// ---------------------------------------------------------------------------

/**
 * Validates the common event envelope fields.
 * Does NOT validate the `payload` — use {@link validateTelemetryEvent} for that.
 */
export function validateEventEnvelope(event: unknown): ValidationResult {
  const errors: ValidationError[] = [];

  if (typeof event !== 'object' || event === null) {
    return {
      valid: false,
      errors: [{ field: 'root', message: 'Event must be a non-null object' }],
    };
  }

  const e = event as Record<string, unknown>;

  if (typeof e.event_id !== 'string' || !UUID_RE.test(e.event_id)) {
    errors.push({ field: 'event_id', message: 'Must be a valid UUID v4 string' });
  }

  if (
    typeof e.event_type !== 'string' ||
    !(VALID_EVENT_TYPES as readonly string[]).includes(e.event_type)
  ) {
    errors.push({
      field: 'event_type',
      message: `Must be one of: ${VALID_EVENT_TYPES.join(', ')}`,
    });
  }

  if (
    typeof e.source !== 'string' ||
    !(VALID_SOURCES as readonly string[]).includes(e.source)
  ) {
    errors.push({
      field: 'source',
      message: `Must be one of: ${VALID_SOURCES.join(', ')}`,
    });
  }

  if (typeof e.timestamp !== 'string' || !ISO_8601_RE.test(e.timestamp)) {
    errors.push({
      field: 'timestamp',
      message:
        'Must be an ISO 8601 UTC timestamp string ending in Z (e.g. 2026-01-01T00:00:00Z)',
    });
  }

  if (
    typeof e.payload !== 'object' ||
    e.payload === null ||
    Array.isArray(e.payload)
  ) {
    errors.push({ field: 'payload', message: 'Must be a non-null object' });
  }

  return { valid: errors.length === 0, errors };
}

// ---------------------------------------------------------------------------
// Payload validators
// ---------------------------------------------------------------------------

/** Validates a `battery_status` payload. */
export function validateBatteryStatusPayload(payload: unknown): ValidationResult {
  const errors: ValidationError[] = [];
  if (typeof payload !== 'object' || payload === null) {
    return { valid: false, errors: [{ field: 'payload', message: 'Must be a non-null object' }] };
  }
  const p = payload as Record<string, unknown>;

  if (
    typeof p.voltage_mv !== 'number' ||
    !Number.isInteger(p.voltage_mv) ||
    p.voltage_mv < 0
  ) {
    errors.push({ field: 'payload.voltage_mv', message: 'Must be a non-negative integer' });
  }

  if (
    typeof p.percentage !== 'number' ||
    p.percentage < 0 ||
    p.percentage > 100
  ) {
    errors.push({ field: 'payload.percentage', message: 'Must be a number between 0 and 100' });
  }

  if (p.voltage_v !== undefined && p.voltage_v !== null) {
    if (typeof p.voltage_v !== 'number' || p.voltage_v < 0) {
      errors.push({ field: 'payload.voltage_v', message: 'Must be a non-negative number' });
    }
  }

  if (p.is_critical !== undefined && p.is_critical !== null) {
    if (typeof p.is_critical !== 'boolean') {
      errors.push({ field: 'payload.is_critical', message: 'Must be a boolean' });
    }
  }

  if (p.battery_type !== undefined && p.battery_type !== null) {
    if (!(VALID_BATTERY_TYPES as readonly string[]).includes(p.battery_type as string)) {
      errors.push({
        field: 'payload.battery_type',
        message: `Must be one of: ${VALID_BATTERY_TYPES.join(', ')}`,
      });
    }
  }

  return { valid: errors.length === 0, errors };
}

/** Validates a `command_received` payload. */
export function validateCommandReceivedPayload(payload: unknown): ValidationResult {
  const errors: ValidationError[] = [];
  if (typeof payload !== 'object' || payload === null) {
    return { valid: false, errors: [{ field: 'payload', message: 'Must be a non-null object' }] };
  }
  const p = payload as Record<string, unknown>;

  if (typeof p.command !== 'string' || p.command.trim() === '') {
    errors.push({ field: 'payload.command', message: 'Must be a non-empty string' });
  }

  if (p.controller_type !== undefined && p.controller_type !== null) {
    if (!(VALID_CONTROLLER_TYPES as readonly string[]).includes(p.controller_type as string)) {
      errors.push({
        field: 'payload.controller_type',
        message: `Must be one of: ${VALID_CONTROLLER_TYPES.join(', ')}`,
      });
    }
  }

  return { valid: errors.length === 0, errors };
}

/** Validates a `command_executed` payload. */
export function validateCommandExecutedPayload(payload: unknown): ValidationResult {
  const errors: ValidationError[] = [];
  if (typeof payload !== 'object' || payload === null) {
    return { valid: false, errors: [{ field: 'payload', message: 'Must be a non-null object' }] };
  }
  const p = payload as Record<string, unknown>;

  if (typeof p.command !== 'string' || p.command.trim() === '') {
    errors.push({ field: 'payload.command', message: 'Must be a non-empty string' });
  }

  if (typeof p.success !== 'boolean') {
    errors.push({ field: 'payload.success', message: 'Must be a boolean' });
  }

  if (p.duration_ms !== undefined && p.duration_ms !== null) {
    if (typeof p.duration_ms !== 'number' || p.duration_ms < 0) {
      errors.push({ field: 'payload.duration_ms', message: 'Must be a non-negative number' });
    }
  }

  return { valid: errors.length === 0, errors };
}

/** Validates a `device_status` payload. */
export function validateDeviceStatusPayload(payload: unknown): ValidationResult {
  const errors: ValidationError[] = [];
  if (typeof payload !== 'object' || payload === null) {
    return { valid: false, errors: [{ field: 'payload', message: 'Must be a non-null object' }] };
  }
  const p = payload as Record<string, unknown>;

  if (typeof p.device_name !== 'string' || p.device_name.trim() === '') {
    errors.push({ field: 'payload.device_name', message: 'Must be a non-empty string' });
  }

  if (
    typeof p.status !== 'string' ||
    !(VALID_DEVICE_STATUSES as readonly string[]).includes(p.status)
  ) {
    errors.push({
      field: 'payload.status',
      message: `Must be one of: ${VALID_DEVICE_STATUSES.join(', ')}`,
    });
  }

  if (p.device_type !== undefined && p.device_type !== null) {
    if (!(VALID_DEVICE_TYPES as readonly string[]).includes(p.device_type as string)) {
      errors.push({
        field: 'payload.device_type',
        message: `Must be one of: ${VALID_DEVICE_TYPES.join(', ')}`,
      });
    }
  }

  return { valid: errors.length === 0, errors };
}

/** Validates an `error` payload. */
export function validateErrorPayload(payload: unknown): ValidationResult {
  const errors: ValidationError[] = [];
  if (typeof payload !== 'object' || payload === null) {
    return { valid: false, errors: [{ field: 'payload', message: 'Must be a non-null object' }] };
  }
  const p = payload as Record<string, unknown>;

  if (typeof p.error_type !== 'string' || p.error_type.trim() === '') {
    errors.push({ field: 'payload.error_type', message: 'Must be a non-empty string' });
  }

  if (typeof p.message !== 'string' || p.message.trim() === '') {
    errors.push({ field: 'payload.message', message: 'Must be a non-empty string' });
  }

  return { valid: errors.length === 0, errors };
}

/** Validates an `api_request` payload. */
export function validateApiRequestPayload(payload: unknown): ValidationResult {
  const errors: ValidationError[] = [];
  if (typeof payload !== 'object' || payload === null) {
    return { valid: false, errors: [{ field: 'payload', message: 'Must be a non-null object' }] };
  }
  const p = payload as Record<string, unknown>;

  if (typeof p.endpoint !== 'string' || p.endpoint.trim() === '') {
    errors.push({ field: 'payload.endpoint', message: 'Must be a non-empty string' });
  }

  if (
    typeof p.status_code !== 'number' ||
    !Number.isInteger(p.status_code) ||
    p.status_code < 100 ||
    p.status_code > 599
  ) {
    errors.push({
      field: 'payload.status_code',
      message: 'Must be an integer HTTP status code (100–599)',
    });
  }

  if (typeof p.latency_ms !== 'number' || p.latency_ms < 0) {
    errors.push({ field: 'payload.latency_ms', message: 'Must be a non-negative number' });
  }

  if (p.method !== undefined && p.method !== null) {
    if (!(VALID_HTTP_METHODS as readonly string[]).includes(p.method as string)) {
      errors.push({
        field: 'payload.method',
        message: `Must be one of: ${VALID_HTTP_METHODS.join(', ')}`,
      });
    }
  }

  return { valid: errors.length === 0, errors };
}

// ---------------------------------------------------------------------------
// Dispatch table
// ---------------------------------------------------------------------------

type PayloadValidator = (payload: unknown) => ValidationResult;

const PAYLOAD_VALIDATORS: Partial<Record<EventType, PayloadValidator>> = {
  battery_status: validateBatteryStatusPayload,
  command_received: validateCommandReceivedPayload,
  command_executed: validateCommandExecutedPayload,
  device_status: validateDeviceStatusPayload,
  error: validateErrorPayload,
  api_request: validateApiRequestPayload,
};

// ---------------------------------------------------------------------------
// Public API
// ---------------------------------------------------------------------------

/**
 * Validates the full event (envelope + payload).
 *
 * @param event - Raw untrusted object from a request body.
 * @returns `{ valid, errors }` where errors is an array of `{ field, message }`.
 */
export function validateTelemetryEvent(event: unknown): ValidationResult {
  const envelopeResult = validateEventEnvelope(event);
  if (!envelopeResult.valid) {
    return envelopeResult;
  }

  const e = event as TelemetryEventEnvelope;
  const payloadValidator = PAYLOAD_VALIDATORS[e.event_type];
  if (!payloadValidator) {
    return { valid: true, errors: [] };
  }

  return payloadValidator(e.payload);
}

/**
 * Validates a batch of events.
 *
 * @param events - Array of raw untrusted objects.
 * @returns Array of `{ index, valid, errors }` for each event.
 */
export function validateTelemetryEventBatch(
  events: unknown[],
): Array<{ index: number; valid: boolean; errors: ValidationError[] }> {
  return events.map((event, index) => ({
    index,
    ...validateTelemetryEvent(event),
  }));
}

/**
 * Convenience guard: returns true if the event is fully valid.
 */
export function isValidTelemetryEvent(event: unknown): event is TelemetryEventEnvelope {
  return validateTelemetryEvent(event).valid;
}

File: robot/controller/telemetry/schemas.py
"""
Wrack telemetry event schemas and validation for the EV3 robot controller.

This module provides:

* String constants for all recognised event types and sources.
* ``validate_event(event)`` — validates the common envelope and the
  event-type-specific payload.  Raises ``ValidationError`` on failure.
* ``validate_payload(event_type, payload)`` — validates just the payload.

The canonical schema definitions live in
``shared/telemetry-types/schemas/*.json``; this module implements the same
constraints in pure Python so it runs on MicroPython without external libs.

When running in a standard CPython environment (e.g. during tests) the module
will optionally use ``jsonschema`` for full Draft-07 compliance.

Usage::

    from robot.controller.telemetry.schemas import validate_event, ValidationError

    try:
        validate_event(event_dict)
    except ValidationError as exc:
        print(exc.errors)
"""

import json
import os
import re

# ``typing`` and ``from __future__ import annotations`` are unavailable on
# Pybricks/MicroPython.  Without the future import, function-signature
# annotations are evaluated at import time, so the fallback below provides a
# subscriptable stub (``Optional[str]`` etc. resolve to the stub harmlessly)
# that lets the module import on the EV3.
try:
    from typing import Any, Dict, List, Optional
except ImportError:  # pragma: no cover - MicroPython runtime path
    class _TypingStub:
        def __getitem__(self, item):
            return self

    Any = Dict = List = Optional = _TypingStub()  # type: ignore[assignment,misc]

# ---------------------------------------------------------------------------
# Public constants
# ---------------------------------------------------------------------------

#: All recognised event sources.
VALID_SOURCES = ["ev3", "rpi", "cloud_functions", "web", "ios"]

#: All recognised event types.
VALID_EVENT_TYPES = [
    "battery_status",
    "command_received",
    "command_executed",
    "device_status",
    "error",
    "api_request",
    "motor_status",
    "sensor_reading",
    "terrain_scan",
    "connection_status",
    "video_stream_start",
    "video_stream_stop",
    "video_stream_health",
]

#: Coarse routing discriminator for the unified ingress (PEN-227): "health"
#: records route to Grafana Cloud, "event" records to BigQuery. Optional on
#: the envelope; absent records default to "event".
VALID_RECORD_TYPES = ["health", "event"]

#: P0-priority event types that have mandatory payload validation.
P0_EVENT_TYPES = [
    "battery_status",
    "command_received",
    "command_executed",
    "device_status",
    "error",
    "api_request",
    "video_stream_start",
    "video_stream_stop",
    "video_stream_health",
]

VALID_DEVICE_STATUSES = [
    "connected", "disconnected", "error", "stalled", "initializing"
]

VALID_DEVICE_TYPES = ["motor", "sensor", "controller", "unknown"]

VALID_BATTERY_TYPES = ["rechargeable", "alkaline", "unknown"]

VALID_CONTROLLER_TYPES = ["ps4", "ps5", "network_remote", "unknown"]

VALID_HTTP_METHODS = ["GET", "POST", "PUT", "PATCH", "DELETE", "OPTIONS"]

# UUID v4 pattern
_UUID_RE = re.compile(
    r"^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$",
    re.IGNORECASE,
)

# ISO 8601 UTC timestamp pattern (YYYY-MM-DDTHH:MM:SS[.fff]Z)
_ISO8601_RE = re.compile(
    r"^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}(\.\d+)?Z$"
)

# ---------------------------------------------------------------------------
# Optional jsonschema integration
# ---------------------------------------------------------------------------

try:
    import jsonschema  # type: ignore[import]
    _JSONSCHEMA_AVAILABLE = True
except ImportError:
    _JSONSCHEMA_AVAILABLE = False


def _load_json_schema(schema_filename: str) -> Optional[Dict[str, Any]]:
    """Load a JSON Schema file from the shared schemas directory."""
    schemas_dir = os.path.join(
        os.path.dirname(__file__),
        "..", "..", "..", "shared", "telemetry-types", "schemas"
    )
    path = os.path.abspath(os.path.join(schemas_dir, schema_filename))
    try:
        with open(path, "r", encoding="utf-8") as fh:
            return json.load(fh)
    except (OSError, json.JSONDecodeError):
        return None


_SCHEMA_FILENAME_MAP = {
    "battery_status": "battery_status.json",
    "command_received": "command_received.json",
    "command_executed": "command_executed.json",
    "device_status": "device_status.json",
    "error": "error.json",
    "api_request": "api_request.json",
    "video_stream_start": "video_stream_start.json",
    "video_stream_stop": "video_stream_stop.json",
    "video_stream_health": "video_stream_health.json",
}

# ---------------------------------------------------------------------------
# Validation error
# ---------------------------------------------------------------------------


class ValidationError(Exception):
    """Raised when an event fails schema validation."""

    def __init__(self, errors: List[str]) -> None:
        self.errors = errors
        super().__init__("; ".join(errors))


# ---------------------------------------------------------------------------
# Envelope validation (pure Python, no external deps)
# ---------------------------------------------------------------------------

def _validate_envelope(event: Any) -> List[str]:
    """Return a list of error messages for the event envelope fields."""
    errors = []

    if not isinstance(event, dict):
        return ["event must be a dict (JSON object)"]

    # event_id
    event_id = event.get("event_id")
    if not isinstance(event_id, str) or not _UUID_RE.match(event_id):
        errors.append("event_id must be a valid UUID v4 string")

    # event_type
    event_type = event.get("event_type")
    if event_type not in VALID_EVENT_TYPES:
        errors.append(
            "event_type must be one of: " + ", ".join(VALID_EVENT_TYPES)
        )

    # source
    source = event.get("source")
    if source not in VALID_SOURCES:
        errors.append("source must be one of: " + ", ".join(VALID_SOURCES))

    # timestamp
    ts = event.get("timestamp")
    if not isinstance(ts, str) or not _ISO8601_RE.match(ts):
        errors.append(
            "timestamp must be an ISO 8601 UTC string ending in Z "
            "(e.g. 2026-01-01T00:00:00Z)"
        )

    # payload
    payload = event.get("payload")
    if not isinstance(payload, dict):
        errors.append("payload must be a dict (JSON object)")

    # type (PEN-227) — optional, routes the event at the unified ingress.
    # VALID_RECORD_TYPES existed as a constant here without ever being
    # checked, so a typo like "heath" passed local validation and only
    # surfaced as a rejection once it reached the ingress — this mirrors
    # the check already present in the TypeScript validators
    # (shared/telemetry-types/typescript/events.ts,
    # cloud/functions/schemas.ts).
    record_type = event.get("type")
    if record_type is not None and record_type not in VALID_RECORD_TYPES:
        errors.append(
            "type must be one of: " + ", ".join(VALID_RECORD_TYPES) + " (or absent)"
        )

    return errors


# ---------------------------------------------------------------------------
# Payload validation (pure Python)
# ---------------------------------------------------------------------------

def _validate_battery_status_payload(payload: Any) -> List[str]:
    errors = []
    if not isinstance(payload, dict):
        return ["battery_status payload must be a dict"]

    voltage_mv = payload.get("voltage_mv")
    if not isinstance(voltage_mv, int) or voltage_mv < 0:
        errors.append("payload.voltage_mv must be a non-negative integer")

    percentage = payload.get("percentage")
    if not isinstance(percentage, (int, float)) or not (0 <= percentage <= 100):
        errors.append("payload.percentage must be a number between 0 and 100")

    if "voltage_v" in payload and payload["voltage_v"] is not None:
        if not isinstance(payload["voltage_v"], (int, float)) or payload["voltage_v"] < 0:
            errors.append("payload.voltage_v must be a non-negative number")

    if "is_critical" in payload and payload["is_critical"] is not None:
        if not isinstance(payload["is_critical"], bool):
            errors.append("payload.is_critical must be a boolean")

    if "battery_type" in payload and payload["battery_type"] is not None:
        if payload["battery_type"] not in VALID_BATTERY_TYPES:
            errors.append(
                "payload.battery_type must be one of: " + ", ".join(VALID_BATTERY_TYPES)
            )

    return errors


def _validate_command_received_payload(payload: Any) -> List[str]:
    errors = []
    if not isinstance(payload, dict):
        return ["command_received payload must be a dict"]

    command = payload.get("command")
    if not isinstance(command, str) or not command.strip():
        errors.append("payload.command must be a non-empty string")

        if "controller_type" in payload and payload["controller_type"] is not None:
            if payload["controller_type"] not in VALID_CONTROLLER_TYPES:
                errors.append(
                    "payload.controller_type must be one of: " + ", ".join(VALID_CONTROLLER_TYPES)
                )

    return errors


def _validate_command_executed_payload(payload: Any) -> List[str]:
    errors = []
    if not isinstance(payload, dict):
        return ["command_executed payload must be a dict"]

    command = payload.get("command")
    if not isinstance(command, str) or not command.strip():
        errors.append("payload.command must be a non-empty string")

    success = payload.get("success")
    if not isinstance(success, bool):
        errors.append("payload.success must be a boolean")

    if "duration_ms" in payload and payload["duration_ms"] is not None:
        duration = payload["duration_ms"]
        if not isinstance(duration, (int, float)) or duration < 0:
            errors.append("payload.duration_ms must be a non-negative number")

    if "controller_type" in payload and payload["controller_type"] is not None:
        if payload["controller_type"] not in VALID_CONTROLLER_TYPES:
            errors.append(
                "payload.controller_type must be one of: " + ", ".join(VALID_CONTROLLER_TYPES)
            )

    return errors


def _validate_device_status_payload(payload: Any) -> List[str]:
    errors = []
    if not isinstance(payload, dict):
        return ["device_status payload must be a dict"]

    device_name = payload.get("device_name")
    if not isinstance(device_name, str) or not device_name.strip():
        errors.append("payload.device_name must be a non-empty string")

    status = payload.get("status")
    if status not in VALID_DEVICE_STATUSES:
        errors.append(
            "payload.status must be one of: " + ", ".join(VALID_DEVICE_STATUSES)
        )

    if "device_type" in payload and payload["device_type"] is not None:
        if payload["device_type"] not in VALID_DEVICE_TYPES:
            errors.append(
                "payload.device_type must be one of: " + ", ".join(VALID_DEVICE_TYPES)
            )

    return errors


def _validate_error_payload(payload: Any) -> List[str]:
    errors = []
    if not isinstance(payload, dict):
        return ["error payload must be a dict"]

    error_type = payload.get("error_type")
    if not isinstance(error_type, str) or not error_type.strip():
        errors.append("payload.error_type must be a non-empty string")

    message = payload.get("message")
    if not isinstance(message, str) or not message.strip():
        errors.append("payload.message must be a non-empty string")

    return errors


def _validate_api_request_payload(payload: Any) -> List[str]:
    errors = []
    if not isinstance(payload, dict):
        return ["api_request payload must be a dict"]

    endpoint = payload.get("endpoint")
    if not isinstance(endpoint, str) or not endpoint.strip():
        errors.append("payload.endpoint must be a non-empty string")

    status_code = payload.get("status_code")
    if not isinstance(status_code, int) or not (100 <= status_code <= 599):
        errors.append("payload.status_code must be an integer between 100 and 599")

    latency_ms = payload.get("latency_ms")
    if not isinstance(latency_ms, (int, float)) or latency_ms < 0:
        errors.append("payload.latency_ms must be a non-negative number")

    if "method" in payload and payload["method"] is not None:
        if payload["method"] not in VALID_HTTP_METHODS:
            errors.append(
                "payload.method must be one of: " + ", ".join(VALID_HTTP_METHODS)
            )

    return errors


VALID_STREAM_PROTOCOLS = ["udp", "tcp", "http"]


def _validate_video_stream_start_payload(payload: Any) -> List[str]:
    errors = []
    if not isinstance(payload, dict):
        return ["video_stream_start payload must be a dict"]

    protocol = payload.get("protocol")
    if protocol not in VALID_STREAM_PROTOCOLS:
            errors.append(
                "payload.protocol must be one of: " + ", ".join(VALID_STREAM_PROTOCOLS)
            )

    port = payload.get("port")
    if not isinstance(port, int) or not (1 <= port <= 65535):
        errors.append("payload.port must be an integer between 1 and 65535")

    for dim in ("resolution_width", "resolution_height"):
        val = payload.get(dim)
        if not isinstance(val, int) or val < 1:
            errors.append("payload." + dim + " must be a positive integer")

    target_fps = payload.get("target_fps")
    if not isinstance(target_fps, (int, float)) or target_fps < 0:
        errors.append("payload.target_fps must be a non-negative number")

    return errors


def _validate_video_stream_stop_payload(payload: Any) -> List[str]:
    errors = []
    if not isinstance(payload, dict):
        return ["video_stream_stop payload must be a dict"]

    reason = payload.get("reason")
    if not isinstance(reason, str) or not reason.strip():
        errors.append("payload.reason must be a non-empty string")

    if "uptime_seconds" in payload and payload["uptime_seconds"] is not None:
        uptime = payload["uptime_seconds"]
        if not isinstance(uptime, (int, float)) or uptime < 0:
            errors.append("payload.uptime_seconds must be a non-negative number")

    return errors


def _validate_video_stream_health_payload(payload: Any) -> List[str]:
    errors = []
    if not isinstance(payload, dict):
        return ["video_stream_health payload must be a dict"]

    fps_recent = payload.get("fps_recent")
    if not isinstance(fps_recent, (int, float)) or fps_recent < 0:
        errors.append("payload.fps_recent must be a non-negative number")

    client_count = payload.get("client_count")
    if not isinstance(client_count, int) or client_count < 0:
        errors.append("payload.client_count must be a non-negative integer")

    frame_drop_total = payload.get("frame_drop_total")
    if not isinstance(frame_drop_total, int) or frame_drop_total < 0:
        errors.append("payload.frame_drop_total must be a non-negative integer")

    uptime_seconds = payload.get("uptime_seconds")
    if not isinstance(uptime_seconds, (int, float)) or uptime_seconds < 0:
        errors.append("payload.uptime_seconds must be a non-negative number")

    return errors


_PAYLOAD_VALIDATORS = {
    "battery_status": _validate_battery_status_payload,
    "command_received": _validate_command_received_payload,
    "command_executed": _validate_command_executed_payload,
    "device_status": _validate_device_status_payload,
    "error": _validate_error_payload,
    "api_request": _validate_api_request_payload,
    "video_stream_start": _validate_video_stream_start_payload,
    "video_stream_stop": _validate_video_stream_stop_payload,
    "video_stream_health": _validate_video_stream_health_payload,
}

# ---------------------------------------------------------------------------
# Public API
# ---------------------------------------------------------------------------


def validate_payload(event_type: str, payload: Any) -> None:
    """
    Validate the payload for a specific event type.

    Raises ``ValidationError`` if the payload is invalid.
    If ``jsonschema`` is installed the validation is additionally checked
    against the canonical JSON Schema file.
    """
    errors = []

    py_validator = _PAYLOAD_VALIDATORS.get(event_type)
    if py_validator:
        errors.extend(py_validator(payload))

    if errors:
        raise ValidationError(errors)

    # Optional: full JSON Schema validation (CPython dev/test environments)
    if _JSONSCHEMA_AVAILABLE:
        schema_file = _SCHEMA_FILENAME_MAP.get(event_type)
        if schema_file:
            schema = _load_json_schema(schema_file)
            if schema is not None:
                # Strip keys whose value is None before validating — JSON Schema
                # Draft-07 does not have a `nullable` keyword; optional fields
                # that are explicitly None are equivalent to being absent.
                clean_payload = {k: v for k, v in payload.items() if v is not None}
                try:
                    jsonschema.validate(instance=clean_payload, schema=schema)
                except jsonschema.ValidationError as exc:
                    raise ValidationError([exc.message]) from exc


def validate_event(event: Any) -> None:
    """
    Validate a complete telemetry event (envelope + payload).

    Raises ``ValidationError`` with a list of error messages if invalid.
    """
    envelope_errors = _validate_envelope(event)
    if envelope_errors:
        raise ValidationError(envelope_errors)

    event_type = event.get("event_type", "")
    payload = event.get("payload", {})
    validate_payload(event_type, payload)


def is_valid_event(event: Any) -> bool:
    """Return True if the event is valid, False otherwise (no exception)."""
    try:
        validate_event(event)
        return True
    except ValidationError:
        return False
